Pydantify framegrab

roxanne-o opened the pull request:
I was motivated to make this change whilst working on SMT. There, we pass around a configuration dictionary that holds the information that framegrab needs. But because the framegrab config object is a dictionary, the user has to know ahead of time all the validation we're doing. This solution is confusing and error prone, especially as we pass around the dictionary and update it in SMT. Additionally the validation logic in Framegrab is uncentralized which makes it really hard to keep track of everything. I wanted to instead in SMT, be passing around an object that we know for certain is validated with the right fields and formats. So I moved all the config logic into Pydantic. Now the config field on a framegrab object is a Pydantic object. There is a specific config model for each framegrab model type, equipped with the custom validation needed for each model. In practice it looks like this:
```
config = RaspberryPiCSI2FrameGrabberConfig(serial_number="1234567890", resolution_width=1280,resolution_height=720, digital_zoom=1.5)
grabber = FrameGrabber.create_grabber(config)

or 

grabber = RaspberryPiCSI2FrameGrabber(config)
```
If the user doesn't want to think about the specific config class they can also use the FramegrabberConfig create method like this:
```
config = FrameGrabberConfig.create(input_type="rpi_csi2", serial_number="1234567890", resolution_width=1280,resolution_height=720, digital_zoom=1.5)
grabber = FrameGrabber.create_grabber(config)
```
This is more similar to the way you would specify config via a dictionary. One key difference is that there is no options field. I think that removing dimensionality from python objects makes things straightforward. The pydantic models however know which fields need to be converted to and from options on the original dictionary object.

Passing in dictionary and yaml configs is still supported with create_grabber/create_grabbers. You can also convert a dictionary config to pydantic config like this
```
config = {
    "name": "Raspberry Pi Ribbon Cable Camera",
    "input_type": "rpi_csi2",
    "options": {
        "resolution": {
            "height": 720,
            "width": 1280
        },
        "zoom": {
            "digital": 1.5
        }
    }
}

pydantic_config = FrameGrabberConfig.from_framegrab_config_dict(config)
```
When I was implementing this, I considered (and implemented :/)  a few other solution. The first was to have the options also be a separate pydantic field like this:
```
camera_options = RTSPOptions(crop={'relative': {'top': 0.1, 'bottom': 0.1}}, digital_zoom=2)
config = FrameGrabberConfig(input_type=InputTypes.RTSP, rtsp_url='rtsp://example.com/stream', options=camera_options)
grabber = create_grabber(config=config)
```
I didn't like that option because there were too many levels of config a user had to create with python. 

I also thought maybe the FrameGrabber objects themselves should be pydantic objects and all of the config could be stored directly on their model like this:
```
grabber=RTSPGrabber(rtsp_url=`rtsp://example.com/stream`, crop={'relative': {'top': 0.1, 'bottom': 0.1}}, digital_zoom=2)
```
This idea is problematic because we do camera start up logic at initialization of the grabber. But if you just want to create configs and pass them around through various processes before actually creating a grabber, it becomes confusing. Because the config itself and its specification is a main ethos of framegrab, I believe it should be its own object.

This still needs to be tested on a realsense and rasperry pi device. I will reach out to people to get one set up for me for testing.

roxanne-o commented on src/framegrab/config.py:
```diff
@@ -0,0 +1,475 @@
+        This function looks for an all-uppercase name between {{ and }} to find an environment
+        variable with that name. If the environment variable is found, its value will be
+        substituted in the rtsp_url.
+        NOTE: This can also work for multiple RTSP URLs in the same config file as long
+            as each one has a unique password placeholder.
```
This note was here before but I don't see how it's true. It looks like multiple RTSP urls can have the same password placeholder, which seems fine to me.

roxanne-o commented on src/framegrab/config.py:
```diff
@@ -0,0 +1,475 @@
+    _unnamed_grabber_count: ClassVar[int] = PrivateAttr(default=0)
+
+    @model_validator(mode="before")
+    def autogenerate_name_if_needed(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        if not values.get("name"):
+            input_type = cls.get_input_type()
+
+            id_field_name = cls.get_input_type_to_id_dict()[input_type]
+            id_field_value = values.get(id_field_name)
+            unnamed_grabber_id = id_field_value if id_field_value else "unnamed"
+
+            unnamed_grabber_count = cls._unnamed_grabber_count
+            values["name"] = f"{input_type.value}_{unnamed_grabber_id}_{unnamed_grabber_count}"
+
+            cls._unnamed_grabber_count += 1
+        return values
```
@timmarkhuff can you carefully review this logic? I don't know enough about why grabbers need unique names in the first place. And in the previous implementations unnamed grabbers weren't even guaranteed unique names: for example for unnamed RTSP urls we gave them the name of the rtsp url. So two grabbers pointed at the same url would have the same name. Same for devices with serial numbers. Is that fine? I don't know why names are important here.

timmarkhuff commented:
I'll dig into this more tomorrow. Looks pretty good at first glance. I did notice one thing though. The grabber configs allow you to provide an input_type. Shouldn't that not be there? The input type is implicit by the type of Pydantic object.

I found that I could do something like this and get a validation error:
```
config = GenericUSBFrameGrabberConfig(
    input_type="usb", # users might try to fill this out, and perhaps make a mistake in doing so
    name="My Camera",
)
```

```
Traceback (most recent call last):
  File "C:\Users\tim\git\framegrab\pr_testing\main.py", line 6, in <module>
    config = GenericUSBFrameGrabberConfig(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\tim\git\framegrab\pr_testing\.venv\Lib\site-packages\pydantic\main.py", line 253, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for GenericUSBFrameGrabberConfig
input_type
  Input should be 'generic_usb', 'rtsp', 'realsense', 'basler', 'rpi_csi2', 'hls', 'youtube_live', 'file' or 'mock' [type=enum, input_value='usb', input_type=str]
    For further information visit https://errors.pydantic.dev/2.11/v/enum
```

Would be better if the user isn't even allowed to provide an `input_type`

timmarkhuff commented:
Setting resolution of generic USB grabbers doesn't work. I tried the following code:
```
from framegrab.config import GenericUSBFrameGrabberConfig
from framegrab import FrameGrabber
import framegrab
from imgcat import imgcat

print(framegrab.__version__)

config = GenericUSBFrameGrabberConfig(
    name="My Camera",
    digital_zoom=1.0,
    num_90_deg_rotations=2,
    resolution_height=1080,
    resolution_width=1920,
)

print(config)

grabber = FrameGrabber.create_grabber(config)

frame = grabber.grab()

print(frame.shape)
imgcat(frame)

grabber.release()
```
And the frame shape prints as (480, 640, 3)

I ran some analogous code in the current version:
```
import framegrab
from imgcat import imgcat

print(framegrab.__version__)

config = {
    "input_type": "generic_usb",
    "options": {
        "resolution": {
            "width": 1920,
            "height": 1080,
        }
    }
}

grabber = framegrab.FrameGrabber.create_grabber(config)
frame = grabber.grab()
print(frame.shape)
imgcat(frame)
grabber.release()
```
The frame shape prints correctly as (1080, 1920, 3)

The reason is that the following line is commented out in `grabbers.py`. 
```
grabber.apply_options(config["options"])
```

roxanne-o commented:
> Setting resolution of generic USB grabbers doesn't work. I tried the following code:
> 
> ```
> from framegrab.config import GenericUSBFrameGrabberConfig
> from framegrab import FrameGrabber
> import framegrab
> from imgcat import imgcat
> 
> print(framegrab.__version__)
> 
> config = GenericUSBFrameGrabberConfig(
>     name="My Camera",
>     digital_zoom=1.0,
>     num_90_deg_rotations=2,
>     resolution_height=1080,
>     resolution_width=1920,
> )
> 
> print(config)
> 
> grabber = FrameGrabber.create_grabber(config)
> 
> frame = grabber.grab()
> 
> print(frame.shape)
> imgcat(frame)
> 
> grabber.release()
> ```
> 
> And the frame shape prints as (480, 640, 3)
> 
> I ran some analogous code in the current version:
> 
> ```
> import framegrab
> from imgcat import imgcat
> 
> print(framegrab.__version__)
> 
> config = {
>     "input_type": "generic_usb",
>     "options": {
>         "resolution": {
>             "width": 1920,
>             "height": 1080,
>         }
>     }
> }
> 
> grabber = framegrab.FrameGrabber.create_grabber(config)
> frame = grabber.grab()
> print(frame.shape)
> imgcat(frame)
> grabber.release()
> ```
> 
> The frame shape prints correctly as (1080, 1920, 3)
> 
> The reason is that the following line is commented out in `grabbers.py`.
> 
> ```
> grabber.apply_options(config["options"])
> ```

Oops nice catch! I fixed the issue. I did have a test for this but I just passed in 480, 640 as the height and width, which is the default I guess? I updated the test to the parameters you set.

timmarkhuff commented:
I tried `framegrab autodiscover` (CLI) and I got this:
```
# Auto-discovered 1 camera(s).
image_sources:
- !!python/object:framegrab.config.GenericUSBFrameGrabberConfig
  __dict__:
    crop: null
    digital_zoom: null
    name: generic_usb_unnamed_0
    num_90_deg_rotations: 0
    resolution_height: null
    resolution_width: null
    serial_number: 7UC20230327V01
  __pydantic_extra__: null
  __pydantic_fields_set__: !!set
    crop: null
    digital_zoom: null
    name: null
    num_90_deg_rotations: null file as YAML:
    serial_number: null
  __pydantic_private__: {}
```

It looks a bit more cluttered than the old version, which looks like this: 
```
# Auto-discovered 1 camera(s).
image_sources:
- id:
    serial_number: 7UC20230327V01
  input_type: generic_usb
  name: GENERIC_USB Camera - 7UC20230327V01
  options: {}
```

Also, more importantly, the auto-generated name in the new (pydantified) version does not include the serial number. The name should be `GENERIC_USB Camera - 7UC20230327V01`. This is important for applications like the URCap that store references to cameras based on name. We want these names to be 1) deterministic, each time you connect to cameras, they should have the same name 2) identifiable: having the serial number is more helpful than just an integer

The logic that generates names like `generic_usb_unnamed_0` without serial numbers is meant to be a fallback when serial numbers aren't available. I think you are right that the logic might have been flawed. It never came into play on the URCap, because we are always able to get serial numbers on linux.

timmarkhuff commented:
> d the issue. I did have a test for this but I just passed in 480, 640 as the height and width, which is the default I guess? I updated the tes

Is your test using a real USB camera? I didn't think we had that. Without a real camera, I'm not sure how we could test that the resolution gets set correctly.

Most cameras will default to a lower resolution like 480p, which is often too low for real applications. Having a test that makes sure we can set 1080p or 720p would be super helpful.

roxanne-o commented:
> generic_usb_unnamed_0

I'm testing with mocks. I do have an actual USB camera, I'm not sure if it's generic though. I think mocks should be okay since we want to test our logic and not the cv2 logic.

roxanne-o commented:
> I tried `framegrab autodiscover` (CLI) and I got this:
> 
> ```
> # Auto-discovered 1 camera(s).
> image_sources:
> - !!python/object:framegrab.config.GenericUSBFrameGrabberConfig
>   __dict__:
>     crop: null
>     digital_zoom: null
>     name: generic_usb_unnamed_0
>     num_90_deg_rotations: 0
>     resolution_height: null
>     resolution_width: null
>     serial_number: 7UC20230327V01
>   __pydantic_extra__: null
>   __pydantic_fields_set__: !!set
>     crop: null
>     digital_zoom: null
>     name: null
>     num_90_deg_rotations: null file as YAML:
>     serial_number: null
>   __pydantic_private__: {}
> ```
> 
> It looks a bit more cluttered than the old version, which looks like this:
> 
> ```
> # Auto-discovered 1 camera(s).
> image_sources:
> - id:
>     serial_number: 7UC20230327V01
>   input_type: generic_usb
>   name: GENERIC_USB Camera - 7UC20230327V01
>   options: {}
> ```
> 
> Also, more importantly, the auto-generated name in the new (pydantified) version does not include the serial number. The name should be `GENERIC_USB Camera - 7UC20230327V01`. This is important for applications like the URCap that store references to cameras based on name. We want these names to be 1) deterministic, each time you connect to cameras, they should have the same name 2) identifiable: having the serial number is more helpful than just an integer
> 
> The logic that generates names like `generic_usb_unnamed_0` without serial numbers is meant to be a fallback when serial numbers aren't available. I think you are right that the logic might have been flawed. It never came into play on the URCap, because we are always able to get serial numbers on linux.

Ah I thought for some reason that since the rtsp_discovery.py didn't call grabber.py then the cli tool wouldn't. I'll update that now.

roxanne-o commented:
> Is your test using a real USB camera? I didn't think we had that. Without a real camera, I'm not sure how we could test that the resolution gets set correctly.
> 
Now tested with a USB camera

> It looks a bit more cluttered than the old version, which looks like this:
> 
Updated autodiscover so now it uses the old format again

> Setting resolution of generic USB grabbers doesn't work.
>
This now works and has been tested with a real camera. I also added a better mock test.

timmarkhuff commented on README.md:
````diff
@@ -72,13 +72,26 @@ lists the sub-commands, including `autodiscover` and `preview`.
+
+or you can directly create the specific grabber config:
+```python
+from framegrab.config import FrameGrabberConfig
````
Should this be `from framegrab.config import RaspberryPiCSI2FrameGrabberConfig`?

timmarkhuff commented on README.md:
````diff
@@ -72,13 +72,26 @@ lists the sub-commands, including `autodiscover` and `preview`.
+or you can directly create the specific grabber config:
+```python
+from framegrab.config import FrameGrabberConfig
+config = RaspberryPiCSI2FrameGrabberConfig(serial_number="1234567890", resolution_width=1280,resolution_height=720, digital_zoom=1.5)
````
I don't believe RaspberryPiCSI2FrameGrabberConfig has a serial_number? In any case i remember that we only support one camera of this type at a time, so specifying a serial number wouldn't do anything useful, even if allowed. Maybe this example should be a generic USB camera instead, since it is so much more common.

roxanne-o replied:
Whoops removed support for specifying a serial number and updated the readme.

timmarkhuff commented on README.md:
````diff
@@ -72,13 +72,26 @@ lists the sub-commands, including `autodiscover` and `preview`.
 
 Here's an example of a single USB camera configured with several options:
 ```python
+from framegrab.config import FrameGrabberConfig
````
I kind of like the second method better. It's what I used when I did my testing. I wonder if we should remove this first method from the documentation to avoid overwhelming people with options? Or maybe at least present it as a secondary option?

This is the method I prefer:
```
from framegrab.config import GenericUSBGrabberConfig
config = GenericUSBGrabberConfig(serial_number="1234567890", resolution_width=1280, resolution_height=720, digital_zoom=1.5)
grabber = FrameGrabber.create_grabber(config)
```

roxanne-o replied:
Let's present it as a secondary option. I found it especially useful for SMT when we might get different kind of grabbers. We can specify that whatever we get goes into some kwargs that go into FrameGrabberConfig.create

timmarkhuff commented on README.md:
```diff
@@ -203,6 +224,474 @@ The table below shows all available configurations and the cameras to which they
 
 In addition to the configurations in the table above, you can set any Basler camera property by including `options.basler.<BASLER PROPERTY NAME>`. For example, it's common to set `options.basler.PixelFormat` to `RGB8`.
 
+#### python FrameGrabberConfig format
```
I don't understand this section. Didn't we already discuss pydantic configurations earlier in the readme?

timmarkhuff commented on write_config_format.py:
```diff
@@ -0,0 +1,44 @@
+"""
+This script generates a YAML string documenting the available parameters for the config of each input type.
```
Cool!

timmarkhuff reviewed (approved):
LGTM. I just added a few comments on the readme.

timmarkhuff commented:
I just found a bug with with basler cameras in this PR.

Run this code...
```
import framegrab
from framegrab.config import BaslerFrameGrabberConfig

print(framegrab.__version__)

config = BaslerFrameGrabberConfig()

grabber = framegrab.FrameGrabber.create_grabber(config)

frame = grabber.grab()

print(frame.shape)

grabber.release()
```
...and you will get this error:
```
  File "/home/tim/git/basler_test/main.py", line 12, in <module>
    grabber = framegrab.FrameGrabber.create_grabber(config)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 281, in create_grabber
    grabber = BaslerFrameGrabber(model_config)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 90, in __init__
    self.apply_options(options)
  File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 905, in apply_options
    for property_name, value in basler_options.items():
                                ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'items'
```
You can work around the issue by passing in an empty dict for basler options:

```
config = BaslerFrameGrabberConfig(basler_options={})
```

roxanne-o commented:
> I just found a bug with with basler cameras in this PR.
> 
> Run this code...
> 
> ```
> import framegrab
> from framegrab.config import BaslerFrameGrabberConfig
> 
> print(framegrab.__version__)
> 
> config = BaslerFrameGrabberConfig()
> 
> grabber = framegrab.FrameGrabber.create_grabber(config)
> 
> frame = grabber.grab()
> 
> print(frame.shape)
> 
> grabber.release()
> ```
> 
> ...and you will get this error:
> 
> ```
>   File "/home/tim/git/basler_test/main.py", line 12, in <module>
>     grabber = framegrab.FrameGrabber.create_grabber(config)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 281, in create_grabber
>     grabber = BaslerFrameGrabber(model_config)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 90, in __init__
>     self.apply_options(options)
>   File "/home/tim/git/basler_test/framegrab/src/framegrab/grabber.py", line 905, in apply_options
>     for property_name, value in basler_options.items():
>                                 ^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'items'
> ```
> 
> You can work around the issue by passing in an empty dict for basler options:
> 
> ```
> config = BaslerFrameGrabberConfig(basler_options={})
> ```

Fixed!
